inst/make_cmdstan.sh

```diff
@@ -28,64 +53,115 @@ if [ -z ${RELDIR} ]; then
    mkdir -p ${RELDIR}
 fi
 
+backup_cmdstan() {
+    if [[ -e cmdstan ]]; then
+        echo "backing up and cleaning installed cmdstan"
+        pushd cmdstan > /dev/null
+        # cleaning the backup
+        if [[ ${WIN} -ne 0 ]]; then
+            mingw32-make clean-all
+        else
+            make clean-all
+        fi
+        pushd ../ > /dev/null
+        # removing existing backup folder
+        if [[ -e cmdstan_backup ]]; then
+            rm -rf cmdstan_backup
+        fi
+        mv cmdstan/ cmdstan_backup/
+    fi
+}
+
+cleanup_cmdstan() {
+    if [[ ${WIN} -ne 0 ]]; then
+        mingw32-make clean-all
+    else
+        make clean-all
+    fi
+}
+
+build_cmdstan() {
+    if [[ ${WIN} -ne 0 ]]; then
+        mingw32-make -j${JOBS} build examples/bernoulli/bernoulli.exe
+    else
+        make -j${JOBS} build examples/bernoulli/bernoulli
+    fi
+}
+
 echo "cmdstan dir: ${RELDIR}"
 
-if [ -z ${VER} ]; then
-    TAG=`curl -s https://api.github.com/repos/stan-dev/cmdstan/releases/latest | grep "tag_name"`
-    echo $TAG > tmp-tag
-    VER=`perl -p -e 's/"tag_name": "v//g; s/",//g' tmp-tag`
-    rm tmp-tag
-fi
-CS=cmdstan-${VER}
-echo "cmdstan version: ${VER}"
+if [[ ${REPO_CLONE} -ne 0 ]]; then
+    pushd ${RELDIR} > /dev/null
+    if [[ ${BACKUP_OLD} -ne 0 ]]; then
+        backup_cmdstan
+    fi
+    git clone --recursive https://github.com/stan-dev/cmdstan.git --single-branch -b develop cmdstan
+else
+    if [ -z ${VER} ]; then
+        TAG=`curl -s https://api.github.com/repos/stan-dev/cmdstan/releases/latest | grep "tag_name"`
+        echo $TAG > tmp-tag
+        VER=`perl -p -e 's/"tag_name": "v//g; s/",//g' tmp-tag`
+        rm tmp-tag
+    fi
 
-pushd ${RELDIR} > /dev/null
-if [[ -d $cs && -f ${CS}/bin/stanc && -f ${CS}/examples/bernoulli/bernoulli ]]; then
-    echo "cmdstan already installed"
-    exit 0
-fi
-echo "installing ${CS}"
-
-echo "downloading ${CS}.tar.gz from github"
-curl -s -OL https://github.com/stan-dev/cmdstan/releases/download/v${VER}/${CS}.tar.gz
-CURL_RC=$?
-if [[ ${CURL_RC} -ne 0 ]]; then
-    echo "github download failed, curl exited with: ${CURL_RC}"
-    exit ${CURL_RC}
-fi
-echo "download complete"
+    CS=cmdstan-${VER}
+    INSTALL_DIR=cmdstan
+    echo "cmdstan version: ${VER}"
 
-echo "unpacking archive"
-if [[ -e cmdstan ]]; then
-    rm -rf cmdstan
-fi
+    pushd ${RELDIR} > /dev/null
+    #testing if there are files in the cmdstan folder
+    if [[ -d ${INSTALL_DIR} && -f ${INSTALL_DIR}/makefile ]]; then
+        echo "cmdstan already installed, checking version"
+        x=$(grep '^CMDSTAN_VERSION := ' ${INSTALL_DIR}/makefile | sed -e 's/CMDSTAN_VERSION := //g')
+        # if the version is the latest dont upgrade
+        # except if forced
+        if [[ ! $VER > $x && $FORCE -ne 1 ]]; then
+            echo "installed cmdstan is the latest version"
+            echo "only running clean and rebuild"
+            pushd cmdstan > /dev/null
+            echo "rebuilding cmdstan binaries"
```

Comment:
If the installed cmdstan is already the latest version I made it clean-all and rebuild. This could be used in case of any weird issue as a first step in troubleshooting.